core/dbt/parser/results.py

```diff
@@ -147,20 +150,29 @@ def _process_node(
         source_file: SourceFile,
         old_file: SourceFile,
         old_result: 'ParseResult',
+        parser_resource_type: NodeType,
     ) -> None:
         """Nodes are a special kind of complicated - there can be multiple
         with the same name, as long as all but one are disabled.
+
+        Only handle nodes where the matching node has the same resource type
+        as the current parser.
         """
-        source_path = source_file.path.original_file_path
+        # cheat: look at the first part of the node ID and compare it to the
+        # parser resource type. On a mismatch, bail out.
+        if parser_resource_type != node_id.split('.')[0]:
+            return
         found: bool = False
         if node_id in old_result.nodes:
             old_node = old_result.nodes[node_id]
-            if old_node.original_file_path == source_path:
+            if old_node.resource_type == parser_resource_type:
```

Comment:
actually, in looking at this further, can you say a little bit more about why we're no longer checking the file path here?